docs/src/main/asciidoc/security-openid-connect-multitenancy-concept.adoc

```diff
@@ -0,0 +1,317 @@
+
+. Provide static configuration for all known tenants and the corresponding OIDC configuration for each of them within the `application.properties` file, or
+. Implement a `@Singleton` CDI bean which implements the `io.quarkus.oidc.TenantConfigResolver` interface. The resolver will be invoked for each inbound call to fetch the OIDC configuration for the active tenant id.
+
+The former approach will become quite cumbersome with increasing number of tenants and environments, so if your application is intended to work with more than few tenants it is recommendable to opt for a more scalable, dynamic approach. Also, it is imperative for enterprises maintaining many multi-tenant applications to use _the same algorithm_ for deriving the tenant identifier and OIDC configuration from inbound requests. This yields the necessary usability and maintainability to enable running a plethora of multi-tenant applications. While technically anything within a service request can be used to find the tenant id, it is practical to *embed the tenant id as a resource path segment within the URL*.
+
+[NOTE]
+====
+**_Question:_** Why is it "practical" to embed the tenant ID as a path segment? Why not as a custom HTTP Header or through other means?
+
+**_(Short) Answer:_** To reduce complexities regarding CORS-aware applications, it is practical to embed the tenant id within the path than most other means.
+
+**_(Longer) Answer:_** During development of multi-tenant applications, it is common to run both frontend and backend services on a development machine - and if so, they have to run on different ports to avoid network clashes. This implies that the browser rendering data from the frontend development application will launch a https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS[CORS OPTIONS preflight call] to the backend service, since the port differs from the front end service. By specification, browsers will not provide custom HTTP headers in the OPTIONS preflight call without special configuration - so if your tenant id is encoded in HTTP Headers, expect problems with required workarounds. Supplying the tenant id as part of the URI resource path largely removes all such problems in the development environment. Moreover, in test or production environments multi-tenant services are often deployed behind network proxies such as Apache or NGinx. Nonstandard HTTP headers require extra configuration in these proxy services to be relayed to the multi-tenant service - but encoding the tenant id within the request path will not such configuration. _You have been warned._
+====
+
+
+The only required method to implement is `Uni<OidcTenantConfig> resolve(RoutingContext routingContext, OidcRequestContext<OidcTenantConfig> requestContext);`. If possible, cache realm IDs and their corresponding `OidcTenantConfig` at application launch to yield good performance during application runtime. A simple structure for resolving tenant configuration is shown below.
```

Comment:
This is a good example but I believe we should clarify here it is a Keycloak specific example and that users can apply a similar technique when working with other providers